src/input-states.js

```diff
@@ -0,0 +1,289 @@
+                delete component.vnis[vnid];
+            }
+        }
+    } else if (port.isMulti()) {
+        return setPortStateArray(component, vnid, portName, state);
+    } else {
+        component.vnis = component.vnis || {};
+        component.vnis[vnid] = component.vnis[vnid] || {};
+        component.vnis[vnid][portName] = state;
+    }
+    return component;
+}
+
+/**
+ * Changes the states of a port.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName name of the port on this component
+ * @param stateArray a list of states for every attached socket on this port
+ */
+function setPortStateArray(component, vnid, portName, stateArray) {
+    var port = component.inPorts[portName];
+    if (!isMultiPort(port))
+        throw Error("This port is not addressable/multi: " + portName);
+    port.listAttached().forEach(function(socketIndex, i) {
+        setPortStateIndex(component, vnid, portName, socketIndex, stateArray[i]);
+    }, []);
+    return component;
+}
+
+/**
+ * Changes the state of a addressable/multi port socket. This is useful for testing
+ * as it looks up the socketIndex from the upstream component name and port.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName name of the port on this component
+ * @param sourceNode the upstream component name
+ * @param sourcePort the upstream component port name attached to this port
+ * @param state the new state of this port socket
+ */
+function setPortStateEntry(component, vnid, portName, sourceNode, sourcePort, state) {
+    if (!isMultiPort(component.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    var socketIndex = findSocketIndex(component, portName, sourceNode, sourcePort);
+    return setPortStateIndex(component, vnid, portName, socketIndex, state);
+}
+
+/**
+ * Changes the state of a addressable/multi port socket. When a new state is received
+ * on a port this should be called.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName name of the port on this component
+ * @param socketIndex the socketIndex (or socketId) pass to the data event handler
+ * @param state the new state of this port socket
+ */
+function setPortStateIndex(component, vnid, portName, socketIndex, state) {
+    if (!isMultiPort(component.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    component.vnis = component.vnis || {};
+    component.vnis[vnid] = component.vnis[vnid] || {};
+    component.vnis[vnid][portName] = component.vnis[vnid][portName] || [];
+    component.vnis[vnid][portName][socketIndex] = state;
+    if (_.isUndefined(state) && !_.compact(component.vnis[vnid][portName]).length) {
+        delete component.vnis[vnid][portName];
```

Comment:
The if condition checks that they are empty. When setting a state to undefined the vnis hash is compacted to avoid memory leaks.